### ISSUE TITLE
Test 80-test_cmp_http fails at alpha17 on NonStop x86/ia64 (round 2)

### ISSUE THREAD

**Issue by rsbeckerca** [openssl/openssl#15571]
The original case was not correctly solved - not sure what was missed in the merge. Please see or reopen #15386.
```
make -j4 -s test TESTS="test_cmp_http"
80-test_cmp_http.t ..
Warning: certificate from 'trusted.crt' with subject '/O=openssl_cmp' is not a CA cert
# cmp_main:/home/git/openssl-3.0/apps/cmp.c:2680:CMP info: using section(s) 'Mock connection' of OpenSSL configuration file '../Mock/test.cnf'
# opt_str:/home/git/openssl-3.0/apps/cmp.c:2203:CMP warning: -proxy option argument is empty string, resetting option
# setup_client_ctx:/home/git/openssl-3.0/apps/cmp.c:1903:CMP info: will contact http://127.0.0.1:80/pkix/
# CMP info: sending IR
# CMP error: connect timeout:
# CMP error: transfer error:request sent: IR, expected response: IP
../../../../util/wrap.pl ../../../../apps/openssl cmp -server '127.0.0.1:0' -config ../Mock/test.cnf -section 'Mock connection' -certout ../../../../test-runs/test_cmp_http/test.cert.pem -proxy '' -no_proxy 127.0.0.1 => 255
    not ok 1 - default config
# ------------------------------------------------------------------------------
    #   Failed test 'default config'
    #   at test/recipes/80-test_cmp_http.t line 147.
Warning: certificate from 'trusted.crt' with subject '/O=openssl_cmp' is not a CA cert
# cmp_main:/home/git/openssl-3.0/apps/cmp.c:2680:CMP info: using section(s) 'Mock connection' of OpenSSL configuration file '../Mock/test.cnf'
# opt_str:/home/git/openssl-3.0/apps/cmp.c:2203:CMP warning: -proxy option argument is empty string, resetting option
# setup_client_ctx:/home/git/openssl-3.0/apps/cmp.c:1903:CMP info: will contact http://example.com:80/pkix/
# CMP info: sending IR
# CMP error: received error:Code=404, Reason=Not Found
# CMP error: error receiving:server=http://example.com:80
# CMP error: transfer error:request sent: IR, expected response: IP
../../../../util/wrap.pl ../../../../apps/openssl cmp -config ../Mock/test.cnf -section 'Mock connection' -certout ../../../../test-runs/test_cmp_http/test.cert.pem -proxy '' -no_proxy 127.0.0.1 -server 'example.com:0' -msg_timeout 1 => 255
    not ok 2 - wrong server
# ------------------------------------------------------------------------------
    #   Failed test 'wrong server'
    #   at test/recipes/80-test_cmp_http.t line 147.
Warning: certificate from 'trusted.crt' with subject '/O=openssl_cmp' is not a CA cert
# cmp_main:/home/git/openssl-3.0/apps/cmp.c:2680:CMP info: using section(s) 'Mock connection' of OpenSSL configuration file '../Mock/test.cnf'
# opt_str:/home/git/openssl-3.0/apps/cmp.c:2203:CMP warning: -proxy option argument is empty string, resetting option
# setup_client_ctx:/home/git/openssl-3.0/apps/cmp.c:1903:CMP info: will contact http://127.0.0.1:99/pkix/
# CMP info: sending IR
# CMP error: connect timeout:
# CMP error: transfer error:request sent: IR, expected response: IP
../../../../util/wrap.pl ../../../../apps/openssl cmp -config ../Mock/test.cnf -section 'Mock connection' -certout ../../../../test-runs/test_cmp_http/test.cert.pem -proxy '' -no_proxy 127.0.0.1 -server '127.0.0.1:99' -msg_timeout 1 => 255
    not ok 3 - wrong server port
# ------------------------------------------------------------------------------
    #   Failed test 'wrong server port'
    #   at test/recipes/80-test_cmp_http.t line 147.
Warning: certificate from 'trusted.crt' with subject '/O=openssl_cmp' is not a CA cert
# cmp_main:/home/git/openssl-3.0/apps/cmp.c:2680:CMP info: using section(s) 'Mock connection' of OpenSSL configuration file '../Mock/test.cnf'
# opt_str:/home/git/openssl-3.0/apps/cmp.c:2203:CMP warning: -proxy option argument is empty string, resetting option
# setup_client_ctx:/home/git/openssl-3.0/apps/cmp.c:1903:CMP info: will contact http://127.0.0.1:80/pkix/
# CMP info: sending IR
# CMP error: connect timeout:
# CMP error: transfer error:request sent: IR, expected response: IP
../../../../util/wrap.pl ../../../../apps/openssl cmp -config ../Mock/test.cnf -section 'Mock connection' -certout ../../../../test-runs/test_cmp_http/test.cert.pem -proxy '' -no_proxy 127.0.0.1 -server 127.0.0.1 -msg_timeout 1 => 255
    not ok 4 - server default port
# ------------------------------------------------------------------------------
    #   Failed test 'server default port'
    #   at test/recipes/80-test_cmp_http.t line 147.
setup_client_ctx:/home/git/openssl-3.0/apps/cmp.c:1783:CMP error: cannot parse -server URL: 127.0.0.1:65536
cmp_main:/home/git/openssl-3.0/apps/cmp.c:2786:CMP error: cannot set up CMP context
# cmp_main:/home/git/openssl-3.0/apps/cmp.c:2680:CMP info: using section(s) 'Mock connection' of OpenSSL configuration file '../Mock/test.cnf'
# opt_str:/home/git/openssl-3.0/apps/cmp.c:2203:CMP warning: -proxy option argument is empty string, resetting option
# CMP error: invalid port number:65536
../../../../util/wrap.pl ../../../../apps/openssl cmp -config ../Mock/test.cnf -section 'Mock connection' -certout ../../../../test-runs/test_cmp_http/test.cert.pem -proxy '' -no_proxy 127.0.0.1 -server '127.0.0.1:65536' => 255
    not ok 5 - server port out of range
# ------------------------------------------------------------------------------
    #   Failed test 'server port out of range'
    #   at test/recipes/80-test_cmp_http.t line 147.
setup_client_ctx:/home/git/openssl-3.0/apps/cmp.c:1783:CMP error: cannot parse -server URL: 127.0.0.1:-10
cmp_main:/home/git/openssl-3.0/apps/cmp.c:2786:CMP error: cannot set up CMP context
# cmp_main:/home/git/openssl-3.0/apps/cmp.c:2680:CMP info: using section(s) 'Mock connection' of OpenSSL configuration file '../Mock/test.cnf'
# opt_str:/home/git/openssl-3.0/apps/cmp.c:2203:CMP warning: -proxy option argument is empty string, resetting option
# CMP error: invalid port number:-10
../../../../util/wrap.pl ../../../../apps/openssl cmp -config ../Mock/test.cnf -section 'Mock connection' -certout ../../../../test-runs/test_cmp_http/test.cert.pem -proxy '' -no_proxy 127.0.0.1 -server '127.0.0.1:-10' => 255
    not ok 6 - server port negative
# ------------------------------------------------------------------------------
    #   Failed test 'server port negative'
    #   at test/recipes/80-test_cmp_http.t line 147.
cmp: Option -server needs a value
cmp: Use -help for summary.
# cmp_main:/home/git/openssl-3.0/apps/cmp.c:2680:CMP info: using section(s) 'Mock connection' of OpenSSL configuration file '../Mock/test.cnf'
# opt_str:/home/git/openssl-3.0/apps/cmp.c:2203:CMP warning: -proxy option argument is empty string, resetting option
../../../../util/wrap.pl ../../../../apps/openssl cmp -config ../Mock/test.cnf -section 'Mock connection' -certout ../../../../test-runs/test_cmp_http/test.cert.pem -proxy '' -no_proxy 127.0.0.1 -server => 255
    not ok 7 - server missing argument
# ------------------------------------------------------------------------------Makefile:3230: recipe for target 'run_tests' failed
make[1]: *** [run_tests] Interrupt
Makefile:3227: recipe for target 'tests' failed
make: *** [tests] Interrupt
```

**Comment by rsbeckerca** [comment]
This may be a duplicate of #15557

**Comment by levitte** [comment]
Please try #15580

**Comment by DDvO** [comment]
Hmm, the first of the test cases should succeed but it does not:
```
# setup_client_ctx:/home/git/openssl-3.0/apps/cmp.c:1903:CMP info: will contact http://127.0.0.1:80/pkix/
# CMP info: sending IR
# CMP error: connect timeout:
```
and this certainly not an issue with `80_test_cmp.t`.

This is likely due to an issue with the BIO IP layer, OS, or compiler.

**Comment by rsbeckerca** [comment]
> Hmm, the first of the test cases should succeed but it does not:
> ```
> # setup_client_ctx:/home/git/openssl-3.0/apps/cmp.c:1903:CMP info: will contact http://127.0.0.1:80/pkix/
> # CMP info: sending IR
> # CMP error: connect timeout:
> ```
> and this certainly not an issue with `80_test_cmp.t`.
> 
> This is likely due to an issue with the BIO IP layer or the NonStop OS.

Why was this working before the merge, but not now? What is different?

**Comment by DDvO** [comment]
> Why was this working before the merge, but not now? What is different?

I have no clue where the discrepancy comes from.
I compared the approved version in https://github.com/siemens/openssl/tree/test_cmp_http_fix_15386_ia64
with the master branch after it got merged - the changes made in the 10 commits (or 5 commits after squashing the fixups) are identical.

**Comment by rsbeckerca** [comment]
> > Why was this working before the merge, but not now? What is different?
> 
> I have no clue where the discrepancy comes from.
> I compared the approved version in https://github.com/siemens/openssl/tree/test_cmp_http_fix_15386_ia64
> with the master branch after it got merged - the changes made in the 10 commits (or 5 commits after squashing the fixups) are identical.

Two thoughts: perhaps /usr/bin/env is misinterpreting `-proxy ''` or the example.com issue (discussed elsewhere) is involved.

**Comment by DDvO** [comment]
> Two thoughts: perhaps /usr/bin/env is misinterpreting `-proxy ''`

The command-line option overrides anything else, and the client does not report to use a proxy, so this is not the issue.

> or the example.com issue (discussed elsewhere) is involved.

Neither - note that in the first test case (and most others as well) the client contacts 127:0.0.1.

Yet here is a clue: the client tries to connect to port 80!

**Comment by DDvO** [comment]
BTW, generally, you can debug the client/server connections via
```
apps/openssl cmp -config "" -srv_ref xyz -srv_secret pass: -rsp_cert test/certs/ee-cert.pem -port 8000 &
apps/openssl cmp -server '127.0.0.1:8000' -config "" -certout test.cert.pem -cmd rr -ref xyz -secret pass: -oldcert test/certs/ee-cert.pem
```
though this will most likely succeed here.
Meanwhile I believe the bug does hide in `80_test_cmp.t`.

**Comment by rsbeckerca** [comment]
> 
> 
> BTW, generally, you can debug the client/server connections via
> 
> ```
> apps/openssl cmp -config "" -srv_ref xyz -srv_secret pass: -rsp_cert test/certs/ee-cert.pem -port 8000 &
> apps/openssl cmp -server '127.0.0.1:8000' -config "" -certout test.cert.pem -cmd rr -ref xyz -secret pass: -oldcert test/certs/ee-cert.pem
> ```
> 
> though this will most likely succeed here.
> Meanwhile I believe the bug does hide in `80_test_cmp.t`.

Exhibit the same problems we saw in #15386:
```
 apps/openssl cmp -config "" -srv_ref xyz -srv_secret pass: -rsp_cert test/certs/ee-cert.pem -port 8000
setup_srv_ctx:/home/git/openssl-3.0/apps/cmp.c:1031:CMP warning: server will not be able to handle signature-protected requests since -srv_trusted is not given
cmp: Error starting accept
CMP error: Address family not supported:calling socket()
CMP error: unable to create socket
CMP error: Address family not supported:calling socket(<NULL>, 8000)
CMP error: unable to create socket
```

**Comment by DDvO** [comment]
Interesting. Did your IP stack change?
I suggest debugging into `acpt_state()` and see if it chooses the right one.

**Comment by rsbeckerca** [comment]
> 
> 
> Interesting. Did your IP stack change?
> I suggest debugging into `acpt_state()` and see if it chooses the right one.

There has been no change in the stack since the tests were done. We did make a change to the DNS servers, but that's not going to impact the test.

**Comment by rsbeckerca** [comment]
> 
> 
> > Interesting. Did your IP stack change?
> > I suggest debugging into `acpt_state()` and see if it chooses the right one.
> 
> There has been no change in the stack since the tests were done. We did make a change to the DNS servers, but that's not going to impact the test.

With that said, I'll get in and debug when the thing builds with @levitte fix.

**Comment by DDvO** [comment]
His fix only changed the Perl test driver, so you can debug this with the master.

**Comment by rsbeckerca** [comment]
> 
> 
> His fix only changed the Perl test driver, so you can debug this with the master.

Too late... I've cleaned and am building - so an hour or two. Doh.

**Comment by DDvO** [comment]
> Too late... I've cleaned and am building - so an hour or two. Doh.

That's a pain.
Even being able to use a proper GNU `make` and `ccache`, 
I tend to keep several copies of `openssl` dirs with the complete Git and build state,
in order to be able to quickly try out things and handle multiple issues in parallel.

**Comment by rsbeckerca** [comment]
> 
> 
> > Too late... I've cleaned and am building - so an hour or two. Doh.
> 
> That's a pain.
> Even being able to use a proper GNU `make` and `ccache`,
> I tend to keep several copies of `openssl` dirs with the complete Git and build state,
> in order to be able to quickly try out things and handle multiple issues in parallel.

`ccache` does not port to NonStop because of non-standard compiler arguments. It would be nice, but really assumes GCC. The requirement to rebuild is partly that the automatic rebuild of the configuration does not work properly on the platform, and without a functional `makedepend`, I have frequent missing dependencies when `libcrypto.so` links.

**Comment by DDvO** [comment]
> The requirement to rebuild is partly that the automatic rebuild of the configuration does not work properly on the platform, and without a functional `makedepend`, I have frequent missing dependencies when `libcrypto.so` links.

I presume you mean that `make depend` does not work properly? What exactly is the problem with the dependencies?
Anyway, as long as the set of source files and the `#include` directives in the sources do not change it should be sufficient to reuse the `Makefile` and any existing object files.

**Comment by rsbeckerca** [comment]
> 
> 
> > The requirement to rebuild is partly that the automatic rebuild of the configuration does not work properly on the platform, and without a functional `makedepend`, I have frequent missing dependencies when `libcrypto.so` links.
> 
> I presume you mean that `make depend` does not work properly? What exactly is the problem with the dependencies?
> Anyway, as long as the set of source files and the `#include` directives in the sources do not change it should be sufficient to reuse the `Makefile` and any existing object files.

I suspect the issue is with the platform `makedepend` rather than anything OpenSSL is doing

**Comment by rsbeckerca** [comment]
The above mentioned fix, the test now hangs. It looks like we are back to where we were 3 weeks ago:
```
00-prep_fipsmodule_cnf.t .. skipped: FIPS module config file only supported in a fips build
Files=1, Tests=0,  1 wallclock secs ( 0.03 usr  0.00 sys +  0.72 cusr  0.00 csys =  0.76 CPU)
Result: NOTESTS
80-test_cmp_http.t ..
    # Looks like you planned 37 tests but ran 1.
not ok 1 - CMP app CLI Mock connection
80-test_cmp_http.t .. 1/?
```
Hangs.

**Comment by DDvO** [comment]
> I suspect the issue is with the platform `makedepend` rather than anything OpenSSL is doing

I see (though I don't know `makedepend`).
How does this surface - what goes wrong if you simply call `make` again to re-build?

**Comment by rsbeckerca** [comment]
> 
> 
> > I suspect the issue is with the platform `makedepend` rather than anything OpenSSL is doing
> 
> I see (though I don't know `makedepend`).
> How does this surface - what goes wrong if you simply call `make` again to re-build?

Honestly, can we stick to the case? I can open an issue separately on the reconfigure, if it becomes an issue - we mostly do clean builds.

The `acpt_state()` seems fine from the `cmp` test. When running with `-port 0`, it also works:
```
$ apps/openssl cmp -config "" -srv_ref xyz -srv_secret pass: -rsp_cert test/certs/ee-cert.pem -port 0
setup_srv_ctx:/home/git/openssl-3.0/apps/cmp.c:1031:CMP warning: server will not be able to handle signature-protected requests since -srv_trusted is not given
ACCEPT 0.0.0.0:15119
```

**Comment by DDvO** [comment]
> The above mentioned fix, the test now hangs. It looks like we are back to where we were 3 weeks ago:
> 
> ```
> 00-prep_fipsmodule_cnf.t .. skipped: FIPS module config file only supported in a fips build
> Files=1, Tests=0,  1 wallclock secs ( 0.03 usr  0.00 sys +  0.72 cusr  0.00 csys =  0.76 CPU)
> Result: NOTESTS
> 80-test_cmp_http.t ..
>     # Looks like you planned 37 tests but ran 1.
> not ok 1 - CMP app CLI Mock connection
> 80-test_cmp_http.t .. 1/?
> ```
> 
> Hangs.

This was expected after the output you gave above.
As I feared already earlier, the ACCEPT reporting recently introduced to `80_cmp_http.t` is not very stable.
I already improved it a little, but this was not sufficient, sigh.
In order to separate such issues from CMP, I introduced an independent test in #15592.

> However, the straight CMP test goes correctly.
> 
> ```
> apps/openssl cmp -config "" -srv_ref xyz -srv_secret pass: -rsp_cert test/certs/ee-cert.pem -port 8000
> setup_srv_ctx:/home/git/openssl-3.0/apps/cmp.c:1031:CMP warning: server will not be able to handle signature-protected requests since -srv_trusted is not given
> ACCEPT 0.0.0.0:8000
> ```
This surprises me, since 3 hours before you reported that
```
 apps/openssl cmp -config "" -srv_ref xyz -srv_secret pass: -rsp_cert test/certs/ee-cert.pem -port 8000
```
failed. What is the difference in the builds (HEAD commits)?

**Comment by rsbeckerca** [comment]
> What is the difference in the builds (HEAD commits)?

```
/home/git/openssl-3.0: git diff --summary fix-15571-15557 origin/master
 delete mode 100644 .github/workflows/cross-compiles.yml
 delete mode 100755 crypto/ec/asm/ecp_nistp521-ppc64.pl
 delete mode 100644 demos/cipher/aesccm.c
 delete mode 100644 demos/cipher/aesgcm.c
 rename demos/{cipher => evp}/Makefile (100%)
 create mode 100644 demos/evp/aesccm.c
 create mode 100644 demos/evp/aesgcm.c
 delete mode 100644 demos/pkey/EVP_PKEY_EC_keygen.c
 delete mode 100644 demos/pkey/Makefile
 rename doc/man3/{EVP_PKEY_get_size.pod => EVP_PKEY_size.pod} (62%)
 delete mode 100644 doc/man3/b2i_PVK_bio_ex.pod
 delete mode 100644 doc/man7/EVP_CIPHER-AES.pod
 delete mode 100644 doc/man7/EVP_CIPHER-ARIA.pod
 delete mode 100644 doc/man7/EVP_CIPHER-BLOWFISH.pod
 delete mode 100644 doc/man7/EVP_CIPHER-CAMELLIA.pod
 delete mode 100644 doc/man7/EVP_CIPHER-CAST.pod
 delete mode 100644 doc/man7/EVP_CIPHER-CHACHA.pod
 delete mode 100644 doc/man7/EVP_CIPHER-DES.pod
 delete mode 100644 doc/man7/EVP_CIPHER-IDEA.pod
 delete mode 100644 doc/man7/EVP_CIPHER-RC2.pod
 delete mode 100644 doc/man7/EVP_CIPHER-RC4.pod
 delete mode 100644 doc/man7/EVP_CIPHER-RC5.pod
 delete mode 100644 doc/man7/EVP_CIPHER-SEED.pod
 delete mode 100644 doc/man7/EVP_CIPHER-SM4.pod
 rename doc/man7/{img => }/kdf.png (100%)
 rename doc/man7/{img => }/mac.png (100%)
 rename doc/man7/{img => }/rand.png (100%)
 delete mode 100644 include/openssl/conftypes.h
 rename providers/common/{include/prov => der}/der_digests.h.in (94%)
 rename providers/common/{include/prov => der}/der_dsa.h.in (95%)
 rename providers/common/{include/prov => der}/der_ec.h.in (95%)
 rename providers/common/{include/prov => der}/der_ecx.h.in (95%)
 rename providers/common/{include/prov => der}/der_rsa.h.in (96%)
 rename providers/common/{include/prov => der}/der_sm2.h.in (95%)
 rename providers/common/{include/prov => der}/der_wrap.h.in (93%)
 delete mode 100644 providers/fips/fips_entry.c
 delete mode 100644 providers/implementations/kdfs/pbkdf1.c
 delete mode 100644 test/certs/dhk2048.pem
 delete mode 100644 test/pbetest.c
 delete mode 100644 test/pkcs7_test.c
 delete mode 100644 test/recipes/00-prep_fipsmodule_cnf.t
 delete mode 100644 test/recipes/05-test_pbe.t
 delete mode 100644 test/recipes/30-test_evp_data/evpkdf_pbkdf1.txt
 delete mode 100644 test/threadstest.h
 delete mode 100644 test/threadstest_fips.c
 delete mode 100644 util/mk-fipsmodule-cnf.pl
```
The test still fails, but `cmp` looks fine in the above command.

**Comment by DDvO** [comment]
Hmm, this (summarized) diff only shows entirely unrelated changes.

**Comment by rsbeckerca** [comment]
> 
> 
> Hmm, this (summarized) diff only shows entirely unrelated changes.

I know. I am completely perplexed and do not know what to look for in the test script.

**Comment by rsbeckerca** [comment]
Some good news, as of commit a4ae661c52:
```
make -j4 -s test TESTS="test_cmp_http"
00-prep_fipsmodule_cnf.t .. skipped: FIPS module config file only supported in a fips build
Files=1, Tests=0,  1 wallclock secs ( 0.03 usr  0.00 sys +  0.71 cusr  0.00 csys =  0.75 CPU)
Result: NOTESTS
80-test_cmp_http.t .. 5/?
#
80-test_cmp_http.t .. ok   d=905969705
All tests successful.
Files=1, Tests=5, 146 wallclock secs ( 0.60 usr  0.00 sys + 54.24 cusr  1.86 csys = 56.70 CPU)
Result: PASS
```

**Comment by DDvO** [comment]
As per https://github.com/openssl/openssl/pull/15580#issuecomment-854914994 we can consider this fixed by #15580 :)

**Comment by rsbeckerca** [comment]
> 
> 
> As per [#15580 (comment)](https://github.com/openssl/openssl/pull/15580#issuecomment-854914994) we can consider this fixed by #15580 :)

Thank you @DDvO @levitte @slontis and others. This had me worried.
